### DISCUSSION TITLE
Relation field doesn't get CollectionId set when using name but it does when using id

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#1428] [Q&A] [answered]
Im on `v0.10.4` I'm creating 2 collections on the `OnBeforeServe()` hook: `listings` and `favourites`

this is how I create the `favourites ` collection **after** `listings` is created :

<details>
  <summary>Code:</summary>
  
```go

if _, err := app.Dao().FindCollectionByNameOrId("favourites"); err != nil {

                        test, _ := app.Dao().FindCollectionByNameOrId("listings")
			collection := &models.Collection{}

			form := forms.NewCollectionUpsert(e.App, collection)
			form.Name = "favourites"
			form.Type = models.CollectionTypeBase
			form.ListRule = types.Pointer("user = @request.auth.id")
			form.ViewRule = types.Pointer("user = @request.auth.id")
			form.CreateRule = types.Pointer("user = @request.auth.id")
			form.UpdateRule = types.Pointer("user = @request.auth.id")
			form.DeleteRule = types.Pointer("user = @request.auth.id")

			form.Schema.AddField(&schema.SchemaField{
				Name:     "user",
				Type:     schema.FieldTypeRelation,
				Required: true,
				Unique:   true,
				Options: &schema.RelationOptions{
					CollectionId:  "_pb_users_auth_",
					CascadeDelete: true,
					MaxSelect:     types.Pointer(1),
				},
			})
			form.Schema.AddField(&schema.SchemaField{
				Name:     "listing",
				Type:     schema.FieldTypeRelation,
				Required: true,
				Unique:   false,
				Options: &schema.RelationOptions{
					MaxSelect:     nil,
					CollectionId:  test.Id,   // CollectionId: "listings", -> this failing
					CascadeDelete: true,
				},
			})

			if err := form.Submit(); err != nil {
				return err
			}
		}
```

</details>

## Issue: 
if I do `CollectionId:  test.Id,` when assigning then the relation collection it works fine, but if I reference the collection by name like `CollectionId:  "listings"`, then something goes wrong because then I go to the admin UI and the collection is not set: 

<details>
  <summary>Screenshot</summary>
  
<img width="661" alt="Screenshot 2022-12-29 at 21 42 39" src="https://user-images.githubusercontent.com/4695692/209962137-a3826dbb-13ff-4328-8f34-df96082c94fb.png">

</details>

Which has unintended effects like the cascade delete option not working (that's how I noticed it). It's weird because there is not error on the debug logs and also if I open directly the db it does appear to be correctly assigned. 

<details>
  <summary>Logs</summary>

  ```
[0.00ms] INSERT INTO `_collections` (`createRule`, `created`, `deleteRule`, `id`, `listRule`, `name`, `options`, `schema`, `system`, `type`, `updateRule`, `updated`, `viewRule`) VALUES ('user = @request.auth.id', '2022-12-29 13:41:58.392Z', 'user = @request.auth.id', 'fkcbjdksa9m917x', 'user = @request.auth.id', 'favourites', '{}', '[{"system":false,"id":"2mkkonnz","name":"user","type":"relation","required":true,"unique":true,"options":{"maxSelect":1,"collectionId":"_pb_users_auth_","cascadeDelete":true}},{"system":false,"id":"vcchcocr","name":"listing","type":"relation","required":true,"unique":false,"options":{"maxSelect":null,"collectionId":"listings","cascadeDelete":true}}]', false, 'base', 'user = @request.auth.id', '2022-12-29 13:41:58.392Z', 'user = @request.auth.id')
[0.00ms] CREATE TABLE `favourites` (`created` TEXT DEFAULT '' NOT NULL, `id` TEXT PRIMARY KEY NOT NULL, `listing` TEXT DEFAULT '', `updated` TEXT DEFAULT '' NOT NULL, `user` TEXT DEFAULT '')
```
</details>

<details>
  <summary>_collections shcema</summary>

id | system | type | name | schema | listRule | viewRule | createRule | updateRule | deleteRule | options | created | updated
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
_pb_users_auth_ | 0 | auth | users | [{"system":false,"id":"users_name","name":"name","type":"text","required":false,"unique":false,"options":{"min":null,"max":null,"pattern":""}},{"system":false,"id":"users_avatar","name":"avatar","type":"file","required":false,"unique":false,"options":{"maxSelect":1,"maxSize":5242880,"mimeTypes":["image/jpg","image/jpeg","image/png","image/svg+xml","image/gif"],"thumbs":null}}] | id = @request.auth.id | id = @request.auth.id |   | id = @request.auth.id | id = @request.auth.id | {"allowEmailAuth":true,"allowOAuth2Auth":true,"allowUsernameAuth":true,"exceptEmailDomains":null,"manageRule":null,"minPasswordLength":8,"onlyEmailDomains":null,"requireEmail":false} | 2022-12-29 13:41:58.388Z | 2022-12-29 13:41:58.388Z
qqsevpmdgnnbfxu | 0 | base | listings | [{"system":false,"id":"nprenzp6","name":"user","type":"relation","required":true,"unique":false,"options":{"maxSelect":1,"collectionId":"users","cascadeDelete":true}},{"system":false,"id":"7cxleirv","name":"title","type":"text","required":true,"unique":false,"options":{"min":null,"max":150,"pattern":""}},{"system":false,"id":"zjggpxck","name":"description","type":"text","required":true,"unique":false,"options":{"min":null,"max":500,"pattern":""}},{"system":false,"id":"ug2i7ra0","name":"price","type":"number","required":true,"unique":false,"options":{"min":null,"max":null}},{"system":false,"id":"kee1tyhw","name":"currecny","type":"text","required":true,"unique":false,"options":{"min":null,"max":null,"pattern":""}},{"system":false,"id":"febwfvvv","name":"photos","type":"file","required":false,"unique":false,"options":{"maxSelect":10,"maxSize":10000000,"mimeTypes":null,"thumbs":null}},{"system":false,"id":"tttoiz9k","name":"videos","type":"file","required":false,"unique":false,"options":{"maxSelect":10,"maxSize":10000000,"mimeTypes":null,"thumbs":null}},{"system":false,"id":"i4bh7ovk","name":"status","type":"text","required":true,"unique":false,"options":{"maxSelect":1,"values":["used","new"]}},{"system":false,"id":"sshudop4","name":"visits","type":"number","required":true,"unique":false,"options":{"min":null,"max":null}},{"system":false,"id":"fixg2rxw","name":"location","type":"text","required":true,"unique":false,"options":{"min":null,"max":null,"pattern":""}},{"system":false,"id":"jjch7nn2","name":"category","type":"text","required":true,"unique":false,"options":{"min":null,"max":null,"pattern":""}}] |   |   | user = @request.auth.id | user = @request.auth.id | user = @request.auth.id | {} | 2022-12-29 13:41:58.390Z | 2022-12-29 13:41:58.390Z
fkcbjdksa9m917x | 0 | base | favourites | [{"system":false,"id":"2mkkonnz","name":"user","type":"relation","required":true,"unique":true,"options":{"maxSelect":1,"collectionId":"_pb_users_auth_","cascadeDelete":true}},{"system":false,"id":"vcchcocr","name":"listing","type":"relation","required":true,"unique":false,"options":{"maxSelect":null,"collectionId":"listings","cascadeDelete":true}}] | user = @request.auth.id | user = @request.auth.id | user = @request.auth.id | user = @request.auth.id | user = @request.auth.id | {} | 2022-12-29 13:41:58.392Z | 2022-12-29 13:41:58.392Z

</details>

**Comment by ganigeorgiev** [comment, marked as answer]
`RelationOptions.CollectionId` must be a collection id.

For backward compatibility reason we allow the collection name in the validator, but it is no longer necessary so I'll consider updating it in a future release.

One thing to note thought. I'm not sure what is your use case but instead of creating the collection in the `OnBeforeServe` event hook you could also consider using migrations - https://pocketbase.io/docs/migrations/.

**Reply by Plorenzo** [reply]
Thanks for the quick answer, I don't know why but I swear I read somewhere that it could be name or id, of course, I can't find it now 🥲 anyway will use id for now on, thanks again! 

I tried using migrations but is not immediately clear to me how to use them to create collections on boot up, the [docs](https://pocketbase.io/docs/collection-methods/#create-new-collection-with-data-validations) for collection creation use `forms.NewCollectionUpsert(app, collection)` which requires an instance of `app` which I don't think is accesible in the migration file.

**Reply by ganigeorgiev** [reply]
>  I swear I read somewhere that it could be name or id, of course, I can't find it now

When requesting the collection or its records with the Web API, for convenience we allow both - id or name, eg:

```
GET /api/collections/:ID/records
// is the same as
GET /api/collections/:NAME/records
```

But that's not the case with `RelationOptions.CollectionId` (or at least not anymore). In earlier PocketBase version this field was named just `Collection` and was used with both id or name, but now we use only the id for the reference because its value is "fixed" and doesn't change.

---

> I tried using migrations but is not immediately clear to me how to use them to create collections on boot up, the [docs](https://pocketbase.io/docs/collection-methods/#create-new-collection-with-data-validations) for collection creation use forms.NewCollectionUpsert(app, collection) which requires an instance of app which I don't think is accesible in the migration file.

That's correct, but usually you don't need validations in a db migration, since there is no external user input to check, and you can use directly the `Dao` (eg. `dao := daos.New(db)`). With v0.9+ if you enable `Automigrate: true`, the required migration Go file will be created automatically for you when making changes from the Admin UI. Alternatively, you can also create a migration with a full snapshot of all your collections by running `migrate collections`. Migrations are applied only once on `serve` or `migrate up`.

**Reply by Plorenzo** [reply]
ok! I'll try again without the data validation and using directly `collection := &models.Collection{}`.

Would be possible to specify the migration order? to be able to get the autogenerated id for the first collection and use it as a field on the second one? something like `listings, _ := daos.New(db).FindCollectionByNameOrId("listings")` I guess or would I need to hardcode some ids?

**Reply by ganigeorgiev** [reply]
> Would be possible to specify the migration order?

The migration files are applied in ASC order based on their filename (they usually start with a timestamp).

> something like listings, _ := daos.New(db).FindCollectionByNameOrId("listings") I guess or would I need to hardcode some ids?

Yes, you should be able to use both approaches; I usually prefer the fixed ids because it is easier to work with and it is less likely to cause migration conflicts on import. The collection id can be any unique string, usually 15 chars, but the length is not required for the collections (with the upcoming multi-match support we'll remove the length requirement for the record models too).
